### ISSUE TITLE
Help with Serial Connection

### ISSUE THREAD

**Issue by dxkruse** [urbanmatrix/mavlink-kotlin#94]
Does this repo support usb serial connections or only UDP/TCP?

**Comment by divyanshupundir** [comment]
@dxkruse, the `connection-core` module is designed so that users can implement their own connections for any platform. The important classes for this are [AbstractMavConnection](https://github.com/divyanshupundir/mavlink-kotlin/blob/main/connections/connection-core/src/commonMain/kotlin/com/divpundir/mavlink/connection/AbstractMavConnection.kt) and [BufferedMavConnection](
https://github.com/divyanshupundir/mavlink-kotlin/blob/main/connections/connection-core/src/commonMain/kotlin/com/divpundir/mavlink/connection/BufferedMavConnection.kt).

As an example, you can have  a look at the [JvmTcpClientMavConnection](https://github.com/divyanshupundir/mavlink-kotlin/blob/main/connections/connection-tcp/src/jvmMain/kotlin/com/divpundir/mavlink/connection/tcp/TcpClientMavConnection.kt) implementation.

In short, your `SerialMavConnection` will need to extend the `AbstractMavConnection` and override the `open()` method. It needs to return a `BufferedMavConnection` which will be created using your serial port.

`BufferedMavConnection` requires `BufferedSource`, `BufferedSink` and a `Closeable` callback to help close the port.

**Comment by dxkruse** [comment]
Awesome, thank you for pointing me in the right direction!

**Comment by divyanshupundir** [comment]
@dxkruse, should I close this issue if you were able to solve the problem?

**Comment by acras01** [comment]
Hello! I've written some code as you pointed. BufferedMavConnection is being created successfully, but when I'm trying to send heartbeat, I get error "com.fazecast.jSerialComm.SerialPortIOException: This port appears to have been shutdown or disconnected."


```
class UsbMavConnection : AbstractMavConnection() {
    override fun open(): MavConnection {
        val ports = SerialPort.getCommPorts().onEach { println(it.descriptivePortName) }

        val serialPort = ports.firstOrNull { it.descriptivePortName.contains(DEVICE_NAME) }
            ?.apply {
                setComPortParameters(BAUD_RATE, 8, SerialPort.ONE_STOP_BIT, SerialPort.NO_PARITY)
                openPort()
            }

        val port = serialPort ?: throw IllegalStateException("Unable to find or open port for $DEVICE_NAME")

        val source = port.inputStream.source().buffer()
        val sink = port.outputStream.sink().buffer()

        return BufferedMavConnection(
            source = source,
            sink = sink,
            resource = { close() },
            dialect = ArdupilotmegaDialect
        ).also { println(it) }
    }
}
```

**Comment by divyanshupundir** [comment]
Hi @acras01, can you share the code snippet of how you are using this class?

**Comment by acras01** [comment]
@divyanshupundir thank you for your respond. problem is fixed. Here is 100% working snippet


```
class UsbMavConnection : AbstractMavConnection() {

    override fun open(): MavConnection {
        val ports = SerialPort.getCommPorts().onEach { println(it.descriptivePortName) }

        val serialPort = ports.firstOrNull { it.descriptivePortName.contains(DEVICE_NAME) }
            ?.apply {
                setComPortParameters(
                    /* newBaudRate = */ BAUD_RATE,
                    /* newDataBits = */ 8,
                    /* newStopBits = */ SerialPort.ONE_STOP_BIT,
                    /* newParity = */ SerialPort.NO_PARITY
                )
                setComPortTimeouts(
                    /* newTimeoutMode = */ SerialPort.TIMEOUT_READ_SEMI_BLOCKING,
                    /* newReadTimeout = */ 0,
                    /* newWriteTimeout = */ 0
                )
                openPort()
            }

        serialPort?.let { port ->
            val source = port.inputStream.source().buffer()
            val sink = port.outputStream.sink().buffer()

            return BufferedMavConnection(
                source = source,
                sink = sink,
                resource = port::closePort,
                dialect = ArdupilotmegaDialect
            )
        } ?: error("Unable to find or open port for $DEVICE_NAME")
    }

    companion object {

        private const val DEVICE_NAME = "Cube Orange+ Mavlink"
        private const val BAUD_RATE = 921600
        private const val DEVICE_VID = "VID_2DAE"
        private const val DEVICE_PID = "PID_1058"
    }
}
```

**Comment by divyanshupundir** [comment]
That's great!

**Comment by nghinguyengr** [comment]
Hi @acras01 , could you please give me on how to configure the Android project to use the SerialPort?

**Comment by acras01** [comment]
Sorry, I used it in desktop project

ср, 7 трав. 2025 р., 06:21 користувач nghinguyen.gr <
***@***.***> пише:

**Comment by nghinguyengr** [comment]
@acras01 Wow, you're using this repo on desktop.
